src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+                if (mappedParameter.HasExplicitDefaultValue && !parameterMapped)
+                {
+                    parameterIsNamed = true;
+                    var generatedExpression = GenerateExpressionFromOptionalParameter(mappedParameter);
+                    var parenthesizedGeneratedExpression = generator.AddParentheses(generatedExpression, includeElasticTrivia: false);
+                    expression = expression.ReplaceNode(oldNode, parenthesizedGeneratedExpression);
+                }
+            }
+
+            return (expression, parameterIsNamed);
+        }
+
+        /// <summary>
+        /// This method goes through all the invocation sites and adds a new argument with the expression to be added.
+        /// It also introduces a parameter at the original method site.
+        /// 
+        /// Example:
+        /// public void M(int x, int y)
+        /// {
+        ///     int f = x * y; // highlight this expression
+        /// }
+        /// 
+        /// public void M1()
+        /// {
+        ///     M(5, 6);
+        /// }
+        /// ---------------------------------------------------->
+        /// 
+        /// public void M(int x, int y, int f)
+        /// {
+        /// }
+        /// 
+        /// public void M1()
+        /// {
+        ///     M(5, 6, 5 * 6);
+        /// }
+        /// </summary>
+        private async Task<Solution> ModifyDocumentInvocationsAndIntroduceParameterAsync(Compilation compilation,
+            Document originalDocument, Document document, Solution modifiedSolution, IMethodSymbol methodSymbol,
+            SyntaxNode containingMethod, TExpressionSyntax expressionCopy, TExpressionSyntax expression,
+            bool allOccurrences, string parameterName,
+            ImmutableArray<TInvocationExpressionSyntax> invocations, CancellationToken cancellationToken)
+        {
+            var mappingDictionary = await MapExpressionToParametersAsync(originalDocument, expressionCopy, cancellationToken).ConfigureAwait(false);
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var editor = new SyntaxEditor(root, generator);
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var insertionIndex = GetInsertionIndex(compilation, methodSymbol, syntaxFacts, containingMethod);
+
+            foreach (var invocationExpression in invocations)
+            {
+                var newArgumentExpression = expression;
+                var invocationArguments = syntaxFacts.GetArgumentsOfInvocationExpression(invocationExpression);
+                var variablesInExpression = expressionCopy.DescendantNodes().OfType<TIdentifierNameSyntax>();
```

Comment:
this is a constant result.  but you invoke it each time through the loop.  hoist out.

Reply:
hoisted to very top caller